File: cole/day06_valid-parenthesis/valid-parenthesis.c
// beats 100% runtime 99.71% memory

typedef unsigned int uint;
#include <stdint.h>

bool isValid(char *s)
{
    uint16_t top = 0;
    for (char *ss = s; *ss; ++ss)
    {
        if (*ss == '(' || *ss == '[' || *ss == '{')
        {
            s[top++] = *ss;
        }
        else
        {
            if (top == 0)
            {
                return false;
            }
            if (*ss == ')' && s[top - 1] != '(')
            {
                return false;
            }
            if (*ss == ']' && s[top - 1] != '[')
            {
                return false;
            }
            if (*ss == '}' && s[top - 1] != '{')
            {
                return false;
            }
            --top;
        }
    }
    return top == 0;
}

File: cole/day06_valid-parenthesis/valid-parenthesis.py
# beats 84%

class Solution:
    def isValid(self, s: str) -> bool:
        l = []
        o = {'{':'}', '(':')', '[':']'}
        c = {'}':'{', ')':'(', ']':'['}
        for ch in s:
            if ch in o:
                l.append(ch)
            if ch in c:
                if not l or l[-1] != c[ch]:
                    return False
                l.pop()
        return not l